synapse/http/server.py

```diff
@@ -353,16 +353,33 @@ def render(self, request):
         """
         Render the request, using an asynchronous render handler if it exists.
         """
-        render_callback_name = "_async_render_" + request.method.decode("ascii")
+        render_callback_name = "render_" + request.method.decode("ascii")
+        async_render_callback_name = "_async_" + render_callback_name
 
-        if hasattr(self, render_callback_name):
-            # Call the handler
-            callback = getattr(self, render_callback_name)
-            defer.ensureDeferred(callback(request))
+        # Try and get the async renderer
+        callback = getattr(self, async_render_callback_name, None)
 
+        # If it doesn't exist, try and get a regular renderer
+        if not callback:
+            callback = getattr(self, render_callback_name, None)
+
+        # No renderer for this request method. Fall back to Twisted's "not
+        # found" handling.
+        if not callback:
+            return super().render(request)
+
+        resp = callback(request)
+
+        # If it's a coroutine, turn it into a Deferred
+        if isinstance(resp, types.CoroutineType):
```

Comment:
wouldn't it be easier to handle the async and and sync cases separately?

(alternatively, if we treat them exactly the same way, why do we bother having two names?)

Reply:
> wouldn't it be easier to handle the async and and sync cases separately?

by which I mean, only do the ensureDeferred stuff in the async case, and only do the `return resp` stuff in the sync case.